draft-ietf-quic-transport.md

```diff
@@ -1618,69 +1618,84 @@ of three ways:
-endpoint MAY send a stateless reset in response to any further incoming packets.
+state.  This results in new packets on the connection being handled generically.
+For instance, an endpoint MAY send a stateless reset in response to any further
+incoming packets.
 
-An endpoint MAY exit the draining period earlier if it can guarantee that its
-peer is also draining.  Receiving a CONNECTION_CLOSE or APPLICATION_CLOSE frame
-is sufficient confirmation, as is receiving a stateless reset.  However,
-disposing of connection state could result in delayed or reordered packets to be
-handled poorly.  For endpoints that have some alternative means to ensure that
-late-arriving packets on the connection do not create QUIC state, such as by
-closing the UDP socket, an abbreviated draining period can allow for faster
-resource recovery.  Servers that retain an open port for accepting new
-connections SHOULD NOT exit the draining period early.
+An endpoint enters a closing period after initiating an immediate close
+({{immediate-close}}) and optionally after an idle timeout ({{idle-timeout}}).
+This state runs for the same duration as the draining period.  While in the
+closing period, an endpoint MUST NOT send packets unless they contain a
+CONNECTION_CLOSE or APPLICATION_CLOSE frame (see {{immediate-close}} for
+details).
 
-The draining period does not apply when a stateless reset ({{stateless-reset}})
-is sent.
+The closing period differs from the draining period only in that closing frames
+can be retransmitted.
+
+An endpoint MAY transition from the closing period to the draining period if it
+can confirm that its peer is also closing or draining.  Receiving a closing
+frame is sufficient confirmation, as is receiving a stateless reset.  The
+draining period SHOULD end when the closing period would have ended.  In other
+words, the endpoint can use the same end time, but cease retransmission of the
+closing packet.
+
+The draining and closing periods do not apply when a stateless reset
+({{stateless-reset}}) is sent.
 
 
 ### Idle Timeout
 
 A connection that remains idle for longer than the idle timeout (see
-{{transport-parameter-definitions}}) becomes closed.  Either peer removes
-connection state if they have neither sent nor received a packet for this time.
+{{transport-parameter-definitions}}) is closed.  At implementation discretion, a
+connection enters either the draining or closing period when the idle timeout
+expires.
 
 The time at which an idle timeout takes effect won't be perfectly synchronized
-on peers.  A connection enters the draining period when the idle timeout
-expires.  During this time, an endpoint that receives new packets MAY choose to
-restore the connection.  Alternatively, an endpoint that receives packets MAY
-signal the timeout using an immediate close.
+on peers.  An endpoint MAY enter a third state where receipt of new packets
+causes the connection to be restored to normal operation.
```

Comment:
"A third state" in a second subsection at the same level feels a little off.  If you want to talk about three closing states, I'd make closing/draining/idle three subsections in one place (or just merge this text into the same section above and don't rely on section headers to demarcate).

Also, these sections no longer line up with the three bullet points immediately before them.  Maybe we need to talk about "closing states" and "ways connections close."

Reply:
I think that I might remove this third state thing and not mention revivification.